Update Go to 1.18

tklauser opened the pull request:
Also switch CI and docs from `go get -u ...` to `go install` as the former is deprecated as of Go 1.17 and no longer supported on Go 1.18 (see https://go.dev/doc/go-get-install-deprecation)

Also change how `golangci-lint` is integrated in the Makefiles and GitHub actions.

See commits for details.

ldelossa reviewed (approved):
approved contingent on tests ✅

rolinh reviewed (changes requested):
Running `go fix ./...` allows to get rid of the now useless `// +build` directives in all the files and only keep the `//go:build` ones.

christarazi reviewed (approved):
:tada:

sayboras reviewed (approved):
Nice 🚀

qmonnet commented on Makefile:
```diff
@@ -473,11 +473,21 @@ govet: ## Run govet on Go source files in the repository.
+bpf-mock-check: ## Check if the helper headers in bpf/mock are up-to-date.
 	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C bpf/mock/ check_helper_headers
 
+lint: golangci-lint bpf-mock-check
```
Can we keep some documentation for the `lint` target? I think the comment after the `##` is parsed and displayed for `make help`.

qmonnet commented on .github/workflows/tests-smoke.yaml:
```diff
@@ -187,8 +188,11 @@ jobs:
           kubectl -n kube-system exec $cilium_pod -- curl http://localhost:9090/metrics > metrics.prom
-          go install github.com/prometheus/prometheus/cmd/promtool@v2.31.1
-          cat metrics.prom | $HOME/go/bin/promtool check metrics
+          curl -sSLO https://github.com/prometheus/prometheus/releases/download/v${PROM_VERSION}/prometheus-${PROM_VERSION}.linux-amd64.tar.gz
```
Should there be a comment that points to the Prometheus issue, so that we can restore simpler steps once it's addressed?

aanm commented on .github/workflows/tests-smoke.yaml:
```diff
@@ -187,7 +187,7 @@ jobs:
           kubectl -n kube-system exec $cilium_pod -- sh -c "apt update && apt install curl -y"
           kubectl -n kube-system exec $cilium_pod -- curl http://localhost:9090/metrics > metrics.prom
-          GO111MODULE=on go get github.com/prometheus/prometheus/cmd/promtool@e4487274853c587717006eeda8804e597d120340 # This is commit hash for v2.24.1
+          go install github.com/prometheus/prometheus/cmd/promtool@v2.31.1
```
Why `2.31.1`? Perhaps we could update to `v2.34.0` and use the pinned commit `881111fec4332c33094a6fb2680c71fffc427275`?

tklauser replied:
This is updated to v2.34.0 as part of the switch to the binary release in the second to last commit. I'll keep the existing version in this commit.

aanm commented on .github/workflows/lint-build-commits.yaml:
```diff
@@ -41,7 +41,7 @@ jobs:
       - name: Install ginkgo
         run: |
-          go get github.com/onsi/ginkgo/ginkgo@v1.12.1
+          go install github.com/onsi/ginkgo/ginkgo@v1.12.1
```
Let's pin the commit sha while we are at it `cc0216944b25a88d3259699a029d4e601fb8a222`

aanm commented on .travis/prepare.sh:
```diff
@@ -45,9 +45,8 @@ install_clang
-# disable go modules to avoid downloading all dependencies when doing go get
-GO111MODULE=off go get golang.org/x/tools/cmd/cover
-GO111MODULE=off go get github.com/mattn/goveralls
+go install golang.org/x/tools/cmd/cover@latest
```
Accordingly the docs, this tools is now part of the std library, should we remove it?

aanm commented on .travis/prepare.sh:
```diff
@@ -45,9 +45,8 @@ install_clang
-GO111MODULE=off go get golang.org/x/tools/cmd/cover
-GO111MODULE=off go get github.com/mattn/goveralls
+go install golang.org/x/tools/cmd/cover@latest
+go install github.com/mattn/goveralls@latest
```
Let's pin it to a commit `a36c7ef8f23b2952fa6e39663f52107dfc8ad69d` [v0.0.11](https://github.com/mattn/goveralls/releases/tag/v0.0.11)

aanm commented on test/consul/gen-cert.sh:
```diff
@@ -18,14 +18,14 @@ gen() {
 		echo "Please install the cfssl utility and make sure you have it in your \$PATH"
 		echo "You can install it in your \$GOPATH by running:"
-		echo "go get -u github.com/cloudflare/cfssl/cmd/cfssl"
+		echo "go install github.com/cloudflare/cfssl/cmd/cfssl@latest"
```
Let's pin to `29ae05fe80e1a9c704ddad7002d90ade7a38cb29` which is `1.6.1`

aanm commented on test/consul/gen-cert.sh:
```diff
@@ -18,14 +18,14 @@ gen() {
 		echo "Please install the cfssljson utility and make sure you have it in your \$PATH"
 		echo "You can install it in your \$GOPATH by running:"
-		echo "go get -u github.com/cloudflare/cfssl/cmd/cfssljson"
+		echo "go install github.com/cloudflare/cfssl/cmd/cfssljson@latest"
```
Let's pin to `29ae05fe80e1a9c704ddad7002d90ade7a38cb29` which is `1.6.1`

aanm commented on test/packet/scripts/install.sh:
```diff
@@ -64,9 +64,8 @@ sudo ln -s /usr/local/go/bin/* /usr/local/bin/
-go get -u github.com/google/gops
-go get -u github.com/onsi/ginkgo/ginkgo
-go get -u github.com/onsi/gomega/...
+go install github.com/google/gops@latest
```
`db6b6e9ba07629c90654a7720c452fb44a864dfb` which is `v0.3.22`

aanm commented on test/packet/scripts/install.sh:
```diff
@@ -64,9 +64,8 @@ sudo ln -s /usr/local/go/bin/* /usr/local/bin/
-go get -u github.com/onsi/ginkgo/ginkgo
-go get -u github.com/onsi/gomega/...
+go install github.com/google/gops@latest
+go install github.com/onsi/ginkgo/ginkgo@latest
```
`1ed6486197ba2a007858c59bdc4d1522631025d5` which is `v2.1.3`

aanm commented on tools/dev-doctor/rootcmd.go:
```diff
@@ -90,7 +90,7 @@ func rootCmdRun(cmd *cobra.Command, args []string) {
 			versionRegexp: regexp.MustCompile(`Ginkgo Version (\d+\.\d+\S*)`),
 			minVersion:    &semver.Version{Major: 1, Minor: 4, Patch: 0},
-			hint:          `Run "go get -u github.com/onsi/ginkgo/ginkgo".`,
+			hint:          `Run "go install github.com/onsi/ginkgo/ginkgo@latest".`,
```
`1ed6486197ba2a007858c59bdc4d1522631025d5`

aanm commented on Makefile:
```diff
@@ -473,11 +473,19 @@ govet: ## Run govet on Go source files in the repository.
 	@$(ECHO_CHECK) golangci-lint
 	$(QUIET) golangci-lint run
+else
+	$(QUIET) $(CONTAINER_ENGINE) run --rm -v `pwd`:/app -w /app docker.io/golangci/golangci-lint:v$(GOLANGCILINT_WANT_VERSION) golangci-lint run
```
`docker.io/golangci/golangci-lint:v$(GOLANGCILINT_WANT_VERSION)@$(GOLANGCILINT_IMAGE_SHA)`

aanm commented on Makefile.defs:
```diff
@@ -64,6 +64,9 @@ GO_TAGS_FLAGS = osusergo
 # RACE is specified.
 GOTEST_COVER_OPTS =
 
+GOLANGCILINT_WANT_VERSION = 1.44.2
```
```
GOLANGCILINT_IMAGE_SHA = sha256:c61216e0aeb8b38dcdcf790e2b7eee6e2f8c5564bd5f938ce09f7a027a409d56
```

(this SHA is for the v1.45.0 so it can be added in the next commit)

rolinh replied:
v1.45.2 is already out.

tklauser replied:
Updated to v1.45.2 (SHA `e84b639c061c8888be91939c78dae9b1525359954e405ab0d9868a46861bd21b`)

aanm commented on .github/workflows/tests-smoke.yaml:
```diff
@@ -187,8 +188,13 @@ jobs:
           kubectl -n kube-system exec $cilium_pod -- curl http://localhost:9090/metrics > metrics.prom
-          go install github.com/prometheus/prometheus/cmd/promtool@v2.31.1
-          cat metrics.prom | $HOME/go/bin/promtool check metrics
+          # Install promtool binary release. `go install` doesn't work due to
```
Small nit, there's a `https://github.com/prometheus/prometheus/releases/download/v2.34.0/sha256sums.txt` that we can use to verify prometheus binaries.

qmonnet reviewed (approved):
Approving again, because I like it. :tophat:

tklauser committed:
CI, docs: use go install instead of go get to install executables

Use `go install` instead of `go get` to install executables. This is the
recommended installation method as of Go 1.17 and is no longer supported
with Go 1.18, see https://go.dev/doc/go-get-install-deprecation

While at it also update the installed version in some cases where
appropriate and drop the installation command for gomega from the
developer docs. Gomega it is not an executable but a library already
vendored in by Cilium.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
.travis: don't install golangci-lint

golanci-lint is run as part of the go-lint GitHub action and the Travis
CI build script isn't invoking any make targets requiring golangci-lint.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
make: run golangci-lint container if not installed

Add a separate golangci-lint target to the Makefile which runs either
the locally installed golangci-lint binary - in case it is found and
matches the requested version - or runs it as a container otherwise.

As part of that, also explicitly require the latest golangci-lint
version v1.45.2.

The same approach is already used in the hubble and cilium-cli repos.

Also add the docker image SHA as suggested by André.

Suggested-by: Robin Hahling <robin.hahling@gw-computing.net>
Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
.github/workflows: update golangci-lint to v1.45.2

That version brings support for Go 1.18.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
.github/workflows: install promtool from binary release

`go get` no longer works to install binaries with Go 1.18 and `go install`
unfortunately doesn't work due to
https://github.com/prometheus/prometheus/issues/8852 and related issues.

Thus install promtool from the Prometheus binary release. While at it
bump the version to v2.34.0.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
Update Go to 1.18

Run `go fix ./...` to remove `// +build` tags.

Fix goleak checks in pkg/hubble/container to work with Go 1.18 due to
change in function on top of stack.

Signed-off-by: Tobias Klauser <tobias@cilium.io>
